### ISSUE TITLE
Cannot filter events in testrpc v2.2.3 using Truffle

### ISSUE THREAD

**Issue by AgeManning** [ConsenSys-archive/ganache-cli-archive#154]
When initialising an event and setting a filter, the returned events do not get filtered. Using the same code on a geth testnet works as expected. 

## Expected Behavior

For a contract containing 

```
event TestEvent(address indexed testAddress, address indexed another, string somethingElse );
```

Using Truffle, I am attempting to get the events using 

```
var filteredEvents =  instance.TestEvent({testAddress: account}, {fromBlock: 0, toBlock: 'latest'});
filteredEvents.get(function(err,logs){console.log(logs);});
```

where `account` has been previously defined. I expect to see only events that have `testAddress = account`, however all events are returned. This behaviour doesn't exist on a private geth testnet. 

## Your Environment

- Version used: v2.2.3
- Environment name and version: Truffle

**Comment by alex-miller-0** [comment]
I am also running into this issue, i.e. filtered event queries are returning all events. I'm using v2.2.4.

**Comment by cryptophonic** [comment]
I have been working on a fix for this.  It's not checked in yet, but try the following patch in the testrpc code.  It seems to be working for me, but I have only a narrow set of use cases.  If this fixes your issues will raise my confidence that it's working correctly.

The patch is built from the following description on how topic filters should work from this page: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_newfilter

**A note on specifying topic filters:**

Topics are order-dependent. A transaction with a log with topics [A, B] will be matched by the following topic filters:

[] "anything"
[A] "A in first position (and anything after)"
[null, B] "anything in first position AND B in second position (and anything after)"
[A, B] "A in first position AND B in second position (and anything after)"
[[A, B], [A, B]] "(A OR B) in first position AND (A OR B) in second position (and anything after)"

```
diff --git a/lib/statemanager.js b/lib/statemanager.js
index 55d5996..1e926a9 100644
--- a/lib/statemanager.js
+++ b/lib/statemanager.js
@@ -393,7 +393,26 @@ StateManager.prototype.getLogs = function(filter, callback) {
       if (err) return finished(err);

       var filtered = blockLogs.filter(function(log) {
-        return expectedAddress == null || log.address == expectedAddress;
+          var match = false;
+          if (expectedAddress === null || log.address === expectedAddress) {
+              match = true;
+              if (Array.isArray(filter.topics)) {
+                  for (var i=0; i<filter.topics.length && i<log.topics.length; i++) {
+                      var innerMatch = false;
+                      if (Array.isArray(filter.topics[i])) {
+                          var or = filter.topics[i];
+                          for (var j=0; j<or.length; j++) {
+                              if (or[j] === log.topics[i])
+                                  innerMatch = true;
+                          }
+                      } else if (filter.topics[i] === null || filter.topics[i] === log.topics[i]) {
+                          innerMatch = true;
+                      }
+                      match = match && innerMatch;
+                  }
+              }
+          }
+          return match;
       });
```

**Comment by tcoulter** [comment]
Thanks for taking a stab at it! This code is O(n^3) (i.e., there are nested loops three deep). Can you try breaking it out at all? I haven't tried it myself but my gut tells me there's a better way.

**Comment by cryptophonic** [comment]
Ok, yes - you're right, it is quite slow.  That's a good place to go next -
the first step was just to try to understand the code and the way testrpc
is built on top of web3-provider-engine, etc..

**Comment by cryptophonic** [comment]
I'm not sure there's much gain in refactoring code but I think what does make sense is breaking out of the inner loops when either the AND condition fails (match == false) or the OR case matches (innerMatch == true).  That would eliminate needless looping.

i.e. something like:

```
      var filtered = blockLogs.filter(function(log) {
          var match = false;
          if (expectedAddress === null || log.address === expectedAddress) {
              match = true;
              if (Array.isArray(filter.topics)) {
                  for (var i=0; i<filter.topics.length && i<log.topics.length; i++) {
                      var innerMatch = false;
                      if (Array.isArray(filter.topics[i])) {
                          var or = filter.topics[i];
                          for (var j=0; j<or.length; j++) {
                              if (or[j] === log.topics[i]) {
                                  innerMatch = true;
                                  break;
                              } 
                          }
                      } else if (filter.topics[i] === null || filter.topics[i] === log.topics[i]) {
                          innerMatch = true;
                      }
                      match = match && innerMatch;
                      if (!match) break;
                  }
              }
          }
          return match;
      });
```

**Comment by tcoulter** [comment]
One thing I'm having an issue with here is getting web3 to send topics to over the RPC at all. i.e., for the following JS:

``` javascript
var event = instance.NumberEvent({number: unexpected_value}, {fromBlock: 0});
```

web3 only makes a request like this:

```
{
  "jsonrpc": "2.0",
  "method": "eth_newFilter",
  "params": [
    {
      "topics": [
        "0xe450ead2824f007ba5fe0defff15b2706ac9d12b679c1cea274df5f4dbf7cc47"
      ],
      "address": "0x3dafdbd46794768d8d75611b26a729eb6c68a90e",
      "fromBlock": "0x0"
    }
  ],
  "id": 61
}
```

If I remove `number: ...` entirely, the request doesn't change:

``` javascript
var event = instance.NumberEvent({}, {fromBlock: 0});
```

produces:

```
{
  "jsonrpc": "2.0",
  "method": "eth_newFilter",
  "params": [
    {
      "topics": [
        "0xe450ead2824f007ba5fe0defff15b2706ac9d12b679c1cea274df5f4dbf7cc47"
      ],
      "address": "0x3dafdbd46794768d8d75611b26a729eb6c68a90e",
      "fromBlock": "0x0"
    }
  ],
  "id": 61
}
```

If web3 doesn't produce the right requests, there's not a whole lot I can do. Has anyone found better documentation than [this](https://github.com/ethereum/wiki/wiki/JavaScript-API#contract-events) regarding contract events?

**Comment by cryptophonic** [comment]
Hi Tim,

Here's some code that works from a browser-based Dapp - I think I remember there was some strangeness until I included _both_ the fromBlock and the toBlock on the second object, i.e. just add a toBlock: "latest" in there.  For example, the filter in the code below works well although I haven't looked at the RPC calls themselves.

API.prototype.getHistory = function(name, from, to, cb) {
    if (cb == null) sync_msg();
    var tickEvent = this.logmgr.MarketTick({mid:name},{fromBlock:from,toBlock:to});
    var arr = [];
    tickEvent.get(function(error, logs) {
        logs.map(function(log) {
            arr.push({
                id: log.transactionHash,
                block: log.blockNumber,
                data: parseFloat(globals.web3.fromWei(log.args.tick))
            });
        });
        cb(arr);
    });
};

**Comment by tcoulter** [comment]
Thanks! Just figured out the issue actually by digging deep into web3.js code. Apparently, the event's parameters need to be indexed if you want to filter on them:

``` javascript
contract EventTest {                
  event NumberEvent(uint indexed number);   

  function triggerEvent(uint val) { 
    NumberEvent(val);              
  }                                 
}
```

Now that I realized that, I'm getting the requests I expect. Will see if I can get this solved.

**Comment by tcoulter** [comment]
Fixed on `develop` branch. Feel free to give it a whirl. Closing this ticket for milestone progress. Will be released with TestRPC 3.0.

Thanks for the feedback all. Appreciate it!
